File: extensions/myna-shell/accent.js
// accent.js — PURE accent-color + reduced-motion resolution (feature
// 004-gnome-shell-indicator, 2026-07-30 wave-ribbon redesign; research R18/
// R19; contract extension.md X25/X26). Resolves GNOME's desktop-wide
// accent-color preference (`org.gnome.desktop.interface`'s `accent-color`,
// GNOME 47+) into the wave ribbon's colour palette, and the `enable-
// animations` preference into a reduced-motion boolean.
//
// Two layers, mirroring the vumeter.js (pure)/dbus.js (Gio-seamed) split
// already established in this bundle:
//   - Pure, directly-testable functions (`resolveAccentPalette`,
//     `resolveReducedMotion`, `derivePalette`) that take already-unwrapped
//     values (a name-or-null, a boolean-or-null) — no Gio import needed to
//     exercise them (test/accent.test.js).
//   - `SystemPreferences`, a small Gio.Settings-backed class (same
//     schema/key-existence-guard + live `changed::` pattern dbus.js uses
//     for org.myna.Dictation) that actually reads the desktop's GSettings
//     and feeds the pure resolvers, with injectable seams for testability.
//
// CRITICAL correctness rule (R18): `Gio.Settings.get_user_value()`, not
// `get_string()`/`St.Settings`, is what lets "the user genuinely chose
// blue" and "the user never touched this setting" (also `'blue'`, GNOME's
// factory default) be told apart — both read identically any other way.

import Gio from 'gi://Gio';

const SCHEMA_ID = 'org.gnome.desktop.interface';
const ACCENT_KEY = 'accent-color';
const MOTION_KEY = 'enable-animations';

// The 9-value libadwaita `Adw.AccentColor` palette (GNOME 47+), from
// libadwaita's `_colors.scss` / `adw_accent_color_to_rgba`.
export const ACCENT_HEX = {
    blue: '#3584e4',
    teal: '#2190a4',
    green: '#3a944a',
    yellow: '#c88800',
    orange: '#ed5b00',
    red: '#e62d42',
    pink: '#d56199',
    purple: '#9141ac',
    slate: '#6f8396',
};

// Fallback when the user has not actively chosen an accent colour
// (untouched default, or the schema/key is unavailable on an older shell) —
// the design decision doc's "default Ubuntu orange".
export const UBUNTU_ORANGE = '#E95420';

// The design decision doc's specific override: when the ribbon's primary
// colour is orange, the darker/complementary secondary tone is a fixed
// Ubuntu-brand aubergine, not a generically-computed colour-wheel
// complement (2026-07-30 /speckit-analyze finding U2 — this had been
// dropped to a generic rule across the derived spec artifacts; reinstated
// here to match the source design brief).
export const UBUNTU_AUBERGINE = '#77216F';

/** Clamp to [0,1]. */
function clamp01(x) {
    return Math.max(0, Math.min(1, x));
}

function hexToRgb(hex) {
    const m = /^#?([0-9a-f]{6})$/i.exec(hex.trim());
    if (!m)
        return {r: 0, g: 0, b: 0};
    const n = parseInt(m[1], 16);
    return {r: (n >> 16) & 0xff, g: (n >> 8) & 0xff, b: n & 0xff};
}

function rgbToHex({r, g, b}) {
    const c = v => Math.round(clamp01(v / 255) * 255).toString(16).padStart(2, '0');
    return `#${c(r)}${c(g)}${c(b)}`;
}

function rgbToHsl({r, g, b}) {
    const rn = r / 255, gn = g / 255, bn = b / 255;
    const max = Math.max(rn, gn, bn), min = Math.min(rn, gn, bn);
    const l = (max + min) / 2;
    if (max === min)
        return {h: 0, s: 0, l};
    const d = max - min;
    const s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
    let h;
    switch (max) {
    case rn: h = (gn - bn) / d + (gn < bn ? 6 : 0); break;
    case gn: h = (bn - rn) / d + 2; break;
    default: h = (rn - gn) / d + 4; break;
    }
    return {h: h / 6, s, l};
}

function hue2rgb(p, q, t) {
    let tt = t;
    if (tt < 0)
        tt += 1;
    if (tt > 1)
        tt -= 1;
    if (tt < 1 / 6)
        return p + (q - p) * 6 * tt;
    if (tt < 1 / 2)
        return q;
    if (tt < 2 / 3)
        return p + (q - p) * (2 / 3 - tt) * 6;
    return p;
}

function hslToRgb({h, s, l}) {
    if (s === 0) {
        const v = Math.round(l * 255);
        return {r: v, g: v, b: v};
    }
    const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
    const p = 2 * l - q;
    return {
        r: Math.round(hue2rgb(p, q, h + 1 / 3) * 255),
        g: Math.round(hue2rgb(p, q, h) * 255),
        b: Math.round(hue2rgb(p, q, h - 1 / 3) * 255),
    };
}

/** Lighten a hex colour toward white by `amount` in [0,1] (the highlight tone). */
function lighten(hex, amount) {
    const {r, g, b} = hexToRgb(hex);
    return rgbToHex({
        r: r + (255 - r) * amount,
        g: g + (255 - g) * amount,
        b: b + (255 - b) * amount,
    });
}

/** True colour-wheel complement (hue rotated 180°), lightness pulled down slightly. */
function complement(hex) {
    const hsl = rgbToHsl(hexToRgb(hex));
    return rgbToHex(hslToRgb({
        h: (hsl.h + 0.5) % 1,
        s: hsl.s,
        l: clamp01(hsl.l * 0.75),
    }));
}

/**
 * Derive the ribbon's full palette from one resolved main colour.
 *
 * @param {string} mainHex - the resolved primary accent hex (`#rrggbb`).
 * @param {boolean} [isOrange] - true when the *name* behind mainHex is
 *     "orange" (either the libadwaita accent or the Ubuntu-orange
 *     fallback) — triggers the aubergine override.
 * @returns {{main: string, highlight: string, darkerComplement: string,
 *     translucentAlpha: number}}
 */
export function derivePalette(mainHex, isOrange = false) {
    return {
        main: mainHex,
        highlight: lighten(mainHex, 0.55),
        darkerComplement: isOrange ? UBUNTU_AUBERGINE : complement(mainHex),
        // The "translucent secondary strand" is the same main colour at
        // reduced alpha, applied by ribbonPaint.js — not a separate hue.
        translucentAlpha: 0.35,
    };
}

/**
 * Resolve the ribbon's palette from the *result* of
 * `Gio.Settings.get_user_value('accent-color')` — a GNOME accent name
 * string if the user has genuinely set one, or `null` if they never have
 * (including sitting on the untouched factory default, itself `'blue'` —
 * R18's core distinction) or the schema/key is unavailable.
 *
 * Never throws: an unrecognized name also falls back to Ubuntu orange.
 *
 * @param {(string|null)} userValue
 * @returns {{main: string, highlight: string, darkerComplement: string,
 *     translucentAlpha: number}}
 */
export function resolveAccentPalette(userValue) {
    if (userValue === null || userValue === undefined)
        return derivePalette(UBUNTU_ORANGE, true);
    const hex = ACCENT_HEX[userValue];
    if (hex === undefined)
        return derivePalette(UBUNTU_ORANGE, true);
    return derivePalette(hex, userValue === 'orange');
}

/**
 * Resolve the reduced-motion boolean from the raw `enable-animations`
 * value (or `null`/`undefined` when the schema/key is unavailable — an
 * older GNOME shell defaults to full motion, per contract X26).
 *
 * @param {(boolean|null)} enableAnimations
 * @returns {boolean} true when the ribbon should render its static/
 *     minimal-motion alternative.
 */
export function resolveReducedMotion(enableAnimations) {
    if (enableAnimations === null || enableAnimations === undefined)
        return false;
    return enableAnimations === false;
}

/**
 * Schema-existence-guarded lookup, never throwing even when the schema is
 * absent (R18). Key existence is reported per key, so a shell with
 * `enable-animations` but no `accent-color` still tracks reduced motion.
 *
 * @param {string} schemaId
 * @param {string[]} keys - keys to probe; those absent are reported so the
 *     caller can skip them rather than dropping the whole schema.
 * @returns {({settings: Gio.Settings, has: object}|null)} the settings
 *     object plus a `has[key]` map, or null if the schema doesn't exist.
 */
function lookupGuardedSettings(schemaId, keys) {
    const source = Gio.SettingsSchemaSource.get_default();
    if (source === null)
        return null;
    const schema = source.lookup(schemaId, true);
    if (schema === null)
        return null;
    const has = {};
    for (const key of keys)
        has[key] = schema.has_key(key);
    return {settings: new Gio.Settings({settings_schema: schema}), has};
}

/**
 * SystemPreferences — the live, Gio-backed half. Reads the desktop's
 * accent-color/enable-animations preferences and re-reads them live via
 * `changed::`, feeding the pure resolvers above. Injectable seam
 * (`_lookupSettings`) mirrors dbus.js's `DictationService` pattern so this
 * is stub-testable the same way, even though the shipped contract tests
 * (T052) exercise the pure resolvers directly rather than this class.
 *
 * Both values are cached and refreshed only from `changed::`. hud.js reads
 * them once per repaint frame, and they change at most a few times a
 * session.
 */
export class SystemPreferences {
    /**
     * @param {object} [callbacks]
     * @param {function((string|null)): void} [callbacks.onAccentChanged]
     * @param {function(boolean): void} [callbacks.onMotionChanged]
     * @param {function(string, string[]): ({settings: Gio.Settings, has: object}|null)} [callbacks._lookupSettings]
     *     test seam (default `lookupGuardedSettings`).
     */
    constructor({
        onAccentChanged = null,
        onMotionChanged = null,
        _lookupSettings = null,
    } = {}) {
        this._onAccentChanged = onAccentChanged ?? (() => {});
        this._onMotionChanged = onMotionChanged ?? (() => {});
        this._lookupSettings = _lookupSettings ?? lookupGuardedSettings;
        this._settings = null;
        this._has = {};
        this._signalIds = [];
        this._palette = resolveAccentPalette(null);
        this._reducedMotion = false;
    }

    /** Current accent palette, safe to call before/after enable(). */
    get accentPalette() {
        return this._palette;
    }

    /** Current reduced-motion boolean, safe to call before/after enable(). */
    get reducedMotion() {
        return this._reducedMotion;
    }

    /** Start watching both keys. Safe to call when the schema is absent. */
    enable() {
        if (this._settings !== null)
            return;
        const found = this._lookupSettings(SCHEMA_ID, [ACCENT_KEY, MOTION_KEY]);
        if (found === null)
            return;
        this._settings = found.settings;
        this._has = found.has ?? {};
        this._refreshAccent();
        this._refreshMotion();
        if (this._has[ACCENT_KEY]) {
            this._signalIds.push(this._settings.connect(
                `changed::${ACCENT_KEY}`, () => {
                    this._refreshAccent();
                    this._onAccentChanged(this._readAccentUserValue());
                }));
        }
        if (this._has[MOTION_KEY]) {
            this._signalIds.push(this._settings.connect(
                `changed::${MOTION_KEY}`, () => {
                    this._refreshMotion();
                    this._onMotionChanged(this._reducedMotion);
                }));
        }
    }

    /** Tear down subscriptions. Safe when dormant. */
    disable() {
        if (this._settings === null)
            return;
        for (const id of this._signalIds)
            this._settings.disconnect(id);
        this._signalIds = [];
        this._settings = null;
        this._has = {};
    }

    _refreshAccent() {
        this._palette = resolveAccentPalette(this._readAccentUserValue());
    }

    _refreshMotion() {
        this._reducedMotion = resolveReducedMotion(this._readMotionValue());
    }

    _readAccentUserValue() {
        if (this._settings === null || !this._has[ACCENT_KEY])
            return null;
        const variant = this._settings.get_user_value(ACCENT_KEY);
        return variant === null ? null : variant.deep_unpack();
    }

    _readMotionValue() {
        if (this._settings === null || !this._has[MOTION_KEY])
            return null;
        return this._settings.get_boolean(MOTION_KEY);
    }
}

File: client/myna-hud/src/accent.rs
//! accent — PURE accent-color resolution (feature 004, 2026-07-30
//! wave-ribbon redesign; research R18, mechanism amended R26; contract
//! extension.md X25). Resolves GNOME's desktop-wide accent-color preference
//! (`org.gnome.desktop.interface`'s `accent-color`) into the wave ribbon's
//! colour palette.
//!
//! The settings-backed half (schema/key-existence-guarded reads, `changed::`
//! subscriptions, the libadwaita style-manager probe) is application-layer
//! wiring; the reduced-motion resolution lives in [`crate::motion`] with
//! E2b's two-safe-sources design.
//!
//! Resolution order (the application layer probes; see
//! [`crate::platform::probe_accent_palette`]):
//!
//! 1. **The theme** — `@accent_bg_color` read back from a styled widget
//!    ([`resolve_theme_accent_palette`]). Direct, needs no version probing,
//!    and correct for Yaru's tints and variants by construction. This is
//!    the analogue of the extension's `-st-accent-color`.
//! 2. `AdwStyleManager:accent-color-rgba` (libadwaita ≥ 1.7) — measured
//!    identical to (1) for every accent.
//! 3. [`fallback_palette`] (Ubuntu orange), only if the desktop reports no
//!    accent whatsoever.
//!
//! **Removed (R18, 2026-08-26)**: the `accent-color` *name* table and the
//! GSettings read that fed it. Both existed to map a settings name onto a
//! colour ourselves, and the theme already answers that question — better,
//! since it also covers Yaru variants and `wartybrown`, which has no
//! upstream enum member. The associated "untouched default is not a
//! choice" rule went with it: its premise (that an untouched Ubuntu
//! desktop reads `'blue'` while looking orange) was false —
//! `ubuntu-settings` ships a gschema override setting
//! `accent-color = 'orange'` — and the rule mis-tinted Yaru accent
//! variants back to plain orange.
//!
use crate::shader::{hex_to_rgb, Rgb, RibbonPalette};

/// The last-resort colour, used only when the desktop reports no accent at
/// all — the design decision doc's "default Ubuntu orange".
pub const UBUNTU_ORANGE: &str = "#E95420";

/// The design decision doc's specific override: when the ribbon's primary
/// colour is orange, the darker/complementary secondary tone is a fixed
/// Ubuntu-brand aubergine, not a generically-computed colour-wheel
/// complement (2026-07-30 /speckit-analyze finding U2 — this had been
/// dropped to a generic rule across the derived spec artifacts; reinstated
/// to match the source design brief).
pub const UBUNTU_AUBERGINE: &str = "#77216F";

/// The resolved ribbon palette (hex strings; convert to the shader's
/// [`RibbonPalette`] with [`AccentPalette::as_ribbon_palette`]).
#[derive(Clone, Debug, PartialEq)]
pub struct AccentPalette {
    pub main: String,
    pub highlight: String,
    pub darker_complement: String,
    /// The "translucent secondary strand" is the same main colour at reduced
    /// alpha, applied by the renderer — not a separate hue.
    pub translucent_alpha: f64,
}

impl AccentPalette {
    /// The main colour as an 0..=1 [`Rgb`] (the shader's uniform space).
    pub fn main_rgb(&self) -> Rgb {
        hex_to_rgb(&self.main)
    }

    /// Convert to the shader module's palette type.
    pub fn as_ribbon_palette(&self) -> RibbonPalette {
        RibbonPalette {
            main: hex_to_rgb(&self.main),
            highlight: hex_to_rgb(&self.highlight),
            darker_complement: hex_to_rgb(&self.darker_complement),
            translucent_alpha: self.translucent_alpha,
        }
    }
}

fn clamp01(x: f64) -> f64 {
    x.clamp(0.0, 1.0)
}

/// Parse `#rrggbb` into 0-255 components; invalid input degrades to black
/// (this module's own rule; non-panicking by design).
fn hex_to_rgb255(hex: &str) -> (u32, u32, u32) {
    let t = hex.trim();
    let h = t.strip_prefix('#').unwrap_or(t);
    if h.len() != 6 || !h.chars().all(|c| c.is_ascii_hexdigit()) {
        return (0, 0, 0);
    }
    let n = u32::from_str_radix(h, 16).unwrap_or(0);
    ((n >> 16) & 0xff, (n >> 8) & 0xff, n & 0xff)
}

fn rgb255_to_hex(r: u32, g: u32, b: u32) -> String {
    let c = |v: u32| format!("{:02x}", (clamp01(v as f64 / 255.0) * 255.0).round() as u32);
    format!("#{}{}{}", c(r), c(g), c(b))
}

fn rgb_to_hsl(r: u32, g: u32, b: u32) -> (f64, f64, f64) {
    let (rn, gn, bn) = (r as f64 / 255.0, g as f64 / 255.0, b as f64 / 255.0);
    let max = rn.max(gn).max(bn);
    let min = rn.min(gn).min(bn);
    let l = (max + min) / 2.0;
    if (max - min).abs() < f64::EPSILON {
        return (0.0, 0.0, l);
    }
    let d = max - min;
    let s = if l > 0.5 {
        d / (2.0 - max - min)
    } else {
        d / (max + min)
    };
    let h = if (max - rn).abs() < f64::EPSILON {
        (gn - bn) / d + if gn < bn { 6.0 } else { 0.0 }
    } else if (max - gn).abs() < f64::EPSILON {
        (bn - rn) / d + 2.0
    } else {
        (rn - gn) / d + 4.0
    };
    (h / 6.0, s, l)
}

fn hue2rgb(p: f64, q: f64, t: f64) -> f64 {
    let t = if t < 0.0 {
        t + 1.0
    } else if t > 1.0 {
        t - 1.0
    } else {
        t
    };
    if t < 1.0 / 6.0 {
        return p + (q - p) * 6.0 * t;
    }
    if t < 1.0 / 2.0 {
        return q;
    }
    if t < 2.0 / 3.0 {
        return p + (q - p) * (2.0 / 3.0 - t) * 6.0;
    }
    p
}

fn hsl_to_rgb255(h: f64, s: f64, l: f64) -> (u32, u32, u32) {
    if s == 0.0 {
        let v = (l * 255.0).round() as u32;
        return (v, v, v);
    }
    let q = if l < 0.5 {
        l * (1.0 + s)
    } else {
        l + s - l * s
    };
    let p = 2.0 * l - q;
    let r = (hue2rgb(p, q, h + 1.0 / 3.0) * 255.0).round() as u32;
    let g = (hue2rgb(p, q, h) * 255.0).round() as u32;
    let b = (hue2rgb(p, q, h - 1.0 / 3.0) * 255.0).round() as u32;
    (r, g, b)
}

/// Lighten a hex colour toward white by `amount` in `[0,1]` (the highlight
/// tone).
fn lighten(hex: &str, amount: f64) -> String {
    let (r, g, b) = hex_to_rgb255(hex);
    rgb255_to_hex(
        (r as f64 + (255.0 - r as f64) * amount).round() as u32,
        (g as f64 + (255.0 - g as f64) * amount).round() as u32,
        (b as f64 + (255.0 - b as f64) * amount).round() as u32,
    )
}

/// True colour-wheel complement (hue rotated 180°), lightness pulled down
/// slightly.
fn complement(hex: &str) -> String {
    let (r, g, b) = hex_to_rgb255(hex);
    let (h, s, l) = rgb_to_hsl(r, g, b);
    let (r, g, b) = hsl_to_rgb255((h + 0.5) % 1.0, s, clamp01(l * 0.75));
    rgb255_to_hex(r, g, b)
}

/// Derive the ribbon's full palette from one resolved main colour.
pub fn derive_palette(main_hex: &str, is_orange: bool) -> AccentPalette {
    AccentPalette {
        main: main_hex.to_string(),
        highlight: lighten(main_hex, 0.55),
        darker_complement: if is_orange {
            UBUNTU_AUBERGINE.to_string()
        } else {
            complement(main_hex)
        },
        translucent_alpha: 0.35,
    }
}

/// The palette used when the desktop reports no accent at all.
pub fn fallback_palette() -> AccentPalette {
    derive_palette(UBUNTU_ORANGE, true)
}

/// The two hexes that count as "orange" for the aubergine override: Yaru's
/// (= [`UBUNTU_ORANGE`]) and upstream libadwaita's.
///
/// Orangeness used to be decided by the settings *name*; resolving the
/// accent from the theme means there is no name to consult, so it is
/// decided by the colour itself.
pub fn is_orange_hex(hex: &str) -> bool {
    let hex = hex.to_ascii_lowercase();
    hex == UBUNTU_ORANGE.to_ascii_lowercase() || hex == "#ed5b00"
}

/// Resolve the palette from the accent **the theme itself reports**
/// (`@accent_bg_color`, read back through
/// [`crate::platform::probe_css_accent`]).
///
/// This is the primary path, and it needs no notion of "did the user
/// choose": the theme reports what the desktop is actually using — Ubuntu
/// orange on an untouched Ubuntu, blue on untouched stock GNOME, a Yaru
/// variant where one is in use, and a deliberate choice where one was
/// made.
pub fn resolve_theme_accent_palette(accent: Rgb) -> AccentPalette {
    let main_hex = rgb255_to_hex(
        (accent.r * 255.0).round() as u32,
        (accent.g * 255.0).round() as u32,
        (accent.b * 255.0).round() as u32,
    );
    let is_orange = is_orange_hex(&main_hex);
    derive_palette(&main_hex, is_orange)
}
